Add batcher signature to Espresso Transaction

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> --> 
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Replaces the Sequencer Signature placeholder the batcher's signature.

Renames `SequencerSignature` to `BatcherSignature` for clarity.

Asana Reference: https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1209445350462276

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209445350462276

Ayiga committed:
Add Batcher Signature to Espresso transaction

In order to be able to verify that a transaction coming from Espresso's finalized
transaction list was originally sourced from the op-batcher, we need a way to
verify that the transaction did actually come from the op-batcher.  To facilitate
this we want the transaction submitted to espresso to be signed by the op-batcher
itself.

The majority of this change is a result of attempting to forward the signing
functions to the place where we need them, while also maintaining no ability to
have direct memory access to any potentially stored private key.

This change signs the payload with the desired private key before it is submitted
to Espresso.

shenkeyao reviewed (changes requested):
Oops as discussed on Zulip, this doesn't seem to be the right key.

shenkeyao commented on op-service/crypto/espresso.go:
```diff
@@ -0,0 +1,143 @@
+	Sign(ctx context.Context, addr common.Address, hash []byte) ([]byte, error)
+}
+
+// SignerFn is a function that signs a transaction with the given address.
```
Nit: Should this comment belong somewhere else? And here we need a comment for `ClientSigner`.

Ayiga replied:
Right you are!  Fixed in bbea92e7399ee3cd0a97767133e89c72e8d4648c

shenkeyao commented on op-service/txmgr/cli.go:
```diff
@@ -346,10 +346,12 @@ func NewConfig(cfg CLIConfig, l log.Logger) (*Config, error) {
 		Backend: l1,
 		ChainID: chainID,
-		Signer:  signerFactory(chainID),
+		Signer:  chainSigner.SignTransaction,
```
Do we still need `Signer` now that `Config` also has `ChainSigner`? Can we get the transaction signer from `ChainSigner` instead?

Ayiga replied:
I think it is possible to get rid of `Signer`, but it would require us modifying the existing test files, `op-service/txmgr/txmgr.go` and a single line in `op-deployer/pkg/deployer/broadcaster/keyed.go`

shenkeyao commented on op-service/crypto/espresso.go:
```diff
@@ -0,0 +1,143 @@
+// ChainSignerFactoryFromConfig considers three ways that signers are created & then creates single factory from those config options.
+// It can either take a remote signer (via opsigner.CLIConfig) or it can be provided either a mnemonic + derivation path or a private key.
+// It prefers the remote signer, then the mnemonic or private key (only one of which can be provided).
+func ChainSignerFactoryFromConfig(l log.Logger, privateKey, mnemonic, hdPath string, signerConfig opsigner.CLIConfig) (ChainSignerFactory, common.Address, error) {
```
Q: Is this function for future testing?

Ayiga replied:
This function is spiritually identical to `SignerFactoryFromConfig`.  It just has a different implementation than the existing function.  This is primarily to ensure that the changes made don't have any larger implications / affects.

`SignerFactoryFromConfig`, only seems to be referenced in a single `op-service/crypto/signature_test.go` file now.  So we could replace it quite easily.

I aimed to have primarily additive changes when making these changes which is why the majority of the changes added are in their own `espresso.go` files within the package.  It should help to minimize the potential conflicts when performing changes.

shenkeyao replied:
Makes sense. I prefer this way of adding the `ChainSigner` version of implementation to `espresso.go` files rather than modifying the existing code, so no need to change!

shenkeyao reviewed (commented):
Looks good overall. Left some questions that may not need changes, but mainly to confirm my understanding.

Ayiga committed:
Update comments on `clientSigner` and `privateKeySigner`

The comment on `privateKeySigner` is for different code altogether.  It should
be replaced with a comment that applies to `privateKeySigner` itself.  Additionally
the comments could use some more explicit purpose.

This change updates the comments on `privateKeySigner` and `clientSigner` that
help to      explain what they are intended for.

Ayiga committed:
Fix ChainSigner not being set for Config and DriverSetup

The ChainSigner is created, but it's never assigned to Config and for DriverSetup.
This leads to null pointer calls when trying to sign the payload for the chain.

shenkeyao commented on op-service/crypto/espresso.go:
```diff
@@ -55,8 +56,10 @@ func (c *clientSigner) SignTransaction(ctx context.Context, address common.Addre
 
-// PrivateKeySignerFn creates a SignerFn that signs transactions with the given
-// private key.
+// privateKeySigner creates a is a ChainSigner that delegates to the stored
```
Nit: Remove "creates a".

Ayiga replied:
Fixed in a817ba4e53dadfd4589b1763272fcfba40941369

shenkeyao reviewed (commented):
Just a nit comment.

Ayiga committed:
Clean up comment that had an extra "creates a"
